From 3ca26c41901f8073509212cdf1c1333e964f8211 Mon Sep 17 00:00:00 2001
From: Will Shanks <willshanks@us.ibm.com>
Date: Fri, 21 Nov 2025 12:59:53 -0500
Subject: [PATCH] Base x86 compiler flags on target rather than host system

`CMAKE_HOST_SYSTEM_PROCESSOR` represents the architecture used to run
cmake while `CMAKE_SYSTEM_PROCESSOR` represents the system that the
project is being compiled for. Previously, the host architecture and a
special OSX variable were used to determine if x86 specific flags should
be used. The OSX variable helped avoid using the x86 build flags for
Apple Silicon cross builds and otherwise x86 was only used to build x86
binaries in CI (aarch64 is built with qemu emulation). However,
cross-building for aarch64 from x86 would fail because the x86 specific
flags would be used. This change uses the target architecture to guard
those flags and removes the OSX variable usage as it is no longer
needed.
---
 CMakeLists.txt                              |  8 +++-----
 qiskit_aer/backends/wrappers/CMakeLists.txt | 10 ++++------
 2 files changed, 7 insertions(+), 11 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2bbda7dc..5e542f82 100755
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -261,12 +261,10 @@ endif()
 
 message(STATUS "BLAS library found: ${BLAS_LIBRARIES}")
 
-if(CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64" OR CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "amd64")
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "amd64")
     if(APPLE OR UNIX)
-        if (NOT CMAKE_OSX_ARCHITECTURES STREQUAL "arm64")
-            set(SIMD_FLAGS_LIST "-mfma;-mavx2")
-	        enable_cxx_compiler_flag_if_supported("-mpopcnt")
-        endif()
+        set(SIMD_FLAGS_LIST "-mfma;-mavx2")
+        enable_cxx_compiler_flag_if_supported("-mpopcnt")
 	elseif(MSVC)
 		set(SIMD_FLAGS_LIST "/arch:AVX2")
 	endif()
diff --git a/qiskit_aer/backends/wrappers/CMakeLists.txt b/qiskit_aer/backends/wrappers/CMakeLists.txt
index c20917fc..c204c385 100644
--- a/qiskit_aer/backends/wrappers/CMakeLists.txt
+++ b/qiskit_aer/backends/wrappers/CMakeLists.txt
@@ -12,12 +12,10 @@ set(AER_LIBRARIES
 	${THRUST_DEPENDANT_LIBS}
 	${MPI_DEPENDANT_LIBS})
 
-if(CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "x86_64" OR CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_HOST_SYSTEM_PROCESSOR STREQUAL "amd64")
-    if (NOT CMAKE_OSX_ARCHITECTURES STREQUAL "arm64")
-      # We build SIMD filed separately, because they will be reached only if the
-      # machine running the code has SIMD support
-      set(SIMD_SOURCE_FILE "${PROJECT_SOURCE_DIR}/src/simulators/statevector/qv_avx2.cpp")
-    endif()
+if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "AMD64" OR CMAKE_SYSTEM_PROCESSOR STREQUAL "amd64")
+    # We build SIMD filed separately, because they will be reached only if the
+    # machine running the code has SIMD support
+    set(SIMD_SOURCE_FILE "${PROJECT_SOURCE_DIR}/src/simulators/statevector/qv_avx2.cpp")
 endif()
 
 set(AER_SIMULATOR_SOURCES "bindings.cc" "${SIMD_SOURCE_FILE}")
-- 
2.49.0

